### PR TITLE
toplevel, show directive for constructors

### PR DESCRIPTION
**Pull request by wyn**
toplevel: `#show Some;;` should show documentation on the option type #8945

I've got it to half work:
<pre>
# #show Some;;
type nonrec 'a <b>Some</b> = None | Some of 'a
</pre>

I think this is because I am passing the Ident of "Some" rather than the "option" into Sig_type - basically I'm not sure how to get from the ident of the value ctor to the ident of the type ctor?

### CODE REVIEW

**Comment by Octachron** [comment]
May I suggest you to edit your title to sum up the object of your PR?
Something like
> [WIP] toplevel, show directive for constructors

Then you could explain what is the issue in the current implementation by editing your first comment.
Otherwise, it takes some time to figure what your PR is about.

File: toplevel/topdirs.ml [hunk 1]
```diff
@@ -549,6 +549,22 @@ let () =
     )
     "Print the signature of the corresponding type constructor."
 
+let () =
+  reg_show_prim "show_constructor"
+    (fun env loc id lid ->
+       let desc = Env.lookup_constructor ~loc Env.Positive lid env in
+       if (Ctype.equal env true [desc.cstr_res] [Predef.type_exn]) then
+         raise Not_found;
```
Comment (Octachron) [inline toplevel/topdirs.ml:571]: You don't need to special case exceptions here, exception constructors are constructors.

File: toplevel/topdirs.ml [hunk 2]
```diff
@@ -549,6 +549,22 @@ let () =
...
+       let path =
+         match Ctype.repr desc.cstr_res with
+         | {desc=Tconstr(path, _, _)} -> path
+         | _ -> raise Not_found
+       in
+       let type_decl = Env.find_type path env in
+       [ Sig_type (id, type_decl, Trec_not, Exported) ]
```
Comment (Octachron) [inline toplevel/topdirs.ml]: Your issue is that you are using the constructor `id` . Since you have the type `path`, if the path is just `Pident id`,  you can use this `Ident.t` directly. Otherwise, you need to synthetize a false `Ident.t` name from this path by concatenating all components of the path.

File: toplevel/topdirs.ml (hunk 1 above)
Comment (alainfrisch) [reply toplevel/topdirs.ml:571]: Their type definition would look like `type exn = ..` which is not very useful. They could be printed as `type exn += Not_found`.

Comment (wyn) [reply toplevel/topdirs.ml:571]: But don't I need to exclude the handling of exceptions in this show_constructors show_prim as exceptions are dealt with in the show _exceptions show_prim - maybe the naming is wrong and it should be "show_non_exception_constructors"

File: toplevel/topdirs.ml (hunk 2 above)
Comment (wyn) [reply toplevel/topdirs.ml]: ah ok, I confess I thought it might be horribly dangerous to make up idents as it seemed to have side-effects (ie incrementing currentstamp) and I didn't know whether this reached further.  I'll push up some changes, thanks for the help.

File: toplevel/topdirs.ml (hunk 1 above)
Comment (wyn) [reply toplevel/topdirs.ml:571]: to be fair there is now some common code that can be put into a little function:
```
let is_exception_constructor env type_expr = 
  Ctype.equal env true [type_expr] [Predef.type_exn]
```

and then used in both show_exception and show_constructor

**Comment by wyn** [comment]
some output:

```
# #show Open_binary;;
type nonrec open_flag =
    Open_rdonly
  | Open_wronly
  | Open_append
  | Open_creat
  | Open_trunc
  | Open_excl
  | Open_binary
  | Open_text
  | Open_nonblock
```
which is the same output as:
```
# #show open_flag;;
type nonrec open_flag =
    Open_rdonly
  | Open_wronly
  | Open_append
  | Open_creat
  | Open_trunc
  | Open_excl
  | Open_binary
  | Open_text
  | Open_nonblock
```

also
```
# #show Some;;
type nonrec 'a option = None | Some of 'a
```

and for exceptions:
```
# #show Failure;;
exception Failure of string
```

File: toplevel/topdirs.ml (hunk 2 above)
Comment (gasche) [reply toplevel/topdirs.ml]: If you have the correct type declaration in `type_decl`, you should be able to use `type_decl.typ_id` to get the identifier of the type.

File: toplevel/topdirs.ml (hunk 1 above)
Comment (gasche) [reply toplevel/topdirs.ml:571]: It makes perfect sense for `#show_exception` toreject non-exception constructors, but I agree that `#show_constructor` should also work on exception constructors, which are "just" a special case of extensible-variant constructors. (It's fine if the printing is not as good as with `#show_exception`.)

Comment (gasche) [reply toplevel/topdirs.ml:571]: I read more of the codebase, and I understand the issue now: the way `#show` is implemented is by calling all the `#show_...` actions in turn, and including in the output the result of each of those that did not raise `Not_found`. For this reason, it is better to raise `Not_found` here, or the output will be duplicated.

@wyn could you add a comment at the raising site that explains this? If we rework the implementation of the generic/agnostic `#show` someday we may be able to lift this limitation.

Comment (gasche) [reply toplevel/topdirs.ml:571]: (One possible approach would be to make `show_constructor` work *well* for exception (print them in a better way than generic extensible variants), by reusing the `show_exception` printing logic, and then simply rebind `show_exception` as an alias `show_constructor` instead of declaring it as a separately-registered show primitive. Then `show_exception` would work on non-exception types, and `show_constructor` would work on exceptions.)

File: toplevel/topdirs.ml (hunk 2 above)
Comment (wyn) [reply toplevel/topdirs.ml]: @gasche it seems that typ_id lives on Typedtree.type_declaration (part of pattern) where-as I am using Env.find_type which returns a Types.type_declaration - which crucially does not have a Ident.t yet.  Not sure if there is an easy way to get from Types to Typedtree, I guess that is what the type checker is doing?

File: toplevel/topdirs.ml (hunk 1 above)
Comment (wyn) [reply toplevel/topdirs.ml:571]: @gasche is that the same thing as having one show_constructor function and just having logic that switches on whether you have an exception or not?

```
let show_constructor bla bla =
  if (check is exception constructor) then
    print out exception stuff
  else
    print out constructor stuff
``` 
That is maybe more easy to understand at the cost of being a slightly longer function

File: toplevel/topdirs.ml (hunk 2 above)
Comment (gasche) [reply toplevel/topdirs.ml]: Ah, my mistake. Thinking about the issue further:

- If the type path is `Foo.t`, it could be nice to show `type Foo.t = ...` (even though this is not valid OCaml declaration syntax), as it shows where the type is declared (I believe that `#show Foo.constructor;;`: in the general case the type may not be available in the current scope). We have the information at hand (it's the path), but the format of returning a valid signature item constrains us a bit; but we can make up a fake identifier with a dot in the name if we want.
- The code of the `show_prim` primitive already makes up a fake identifier: `let id = Ident.create_persistent s in ...`. This was enough for most need, but not this one where we want to talk about another identifier than the one corresponding to the Longident.t given by the user. So we could reuse the same hack and use `Ident.create_persistent (Path.name path)`.

**Comment by gasche** [comment]
@wyn could you include some testcases for the new behavior in the compiler testsuite?

Right now there is no testsuite for the `#show` directive(s)`. Let me walk you through the process of adding one.

First, create a file `testsuite/tests/tool-toplevel/show.ml` with the following content (testing for existing features of `#show`):

```ocaml
(* TEST
   * expect 
*)

module type S = sig type t val x : t end;;
module M : S = struct type t = int let x = 3 end;;
[%%expect {|
|}];;

#show M;;
[%%expect {|
|}];;

#show S;;
[%%expect {|
|}];;

#show Invalid_argument;;
[%%expect {|
|}];;

#show option;;
[%%expect {|
|}];;
```

(the `expect` blocks should contain the expected output, but I don't write them by hand, I use the testsuite to fill them out)

Then, run the tests on the file. There are two ways to do it, from the `testsuite/` directory:

```
../ocamltest/ocamltest.opt tests/tool-toplevel/show.ml
# or
make one DIR=tests/tool-toplevel # (runs all tests in the directory, not just this one)
```

The tests will fail because the `expect` sections are empty. At this point I usually `git add` the file, to store the reference version, and then I run "promote" to fill the expect-test blocks:

```
../ocamltest/ocamltest.opt -promote tests/tool-toplevel/show.ml
# or
make promote DIR=tests/tool-toplevel
```

and then `git diff` contains exactly the changes made by `promote`. I check that they are correct (as expected), and now there is a testfile ready to commit.

One you have this initial file and the promotion process, it should be easy to add new tests. Tests are good for correctness, to specify corner cases, but they can also serve as simple examples of how your feature work (all the examples you have shown in this PR discussion can go into tests).

File: toplevel/topdirs.ml (hunk 1 above)
Comment (gasche) [reply toplevel/topdirs.ml:571]: Yes, this structure is fine. Note that this discussion turns out to be about a choice between two behaviors:

1. either `show_constructor` fails on exceptions, and `show_exception` supports only exceptions
2. or `show_constructor` works on exceptions, and `show_exception` (now simply an alias of `show_constructor`) supports exceptions and also non-exception constructors

None of those options are fully satisfying, but they work well with the current "try everything and aggregate the result" approach of the `#show` implementation. I'm not completely sure which one to prefer. If you are also not sure, then keeping the code as you have written it currently (with a comment at the raise point to explain why it's important to fail on exception constructors) is also a perfectly good option.

**Comment by wyn** [comment]
sure thing, I'll do that next. 

I'd also be happy to  create a bunch of new issues for the other things people have suggested if that would be helpful? 

For the final PR do people usually squash the commits into one complete change?

**Comment by Octachron** [comment]
People often clean the history to keep only "meaningful" commits. Note that you don't need to open a new PR, just removing the WIP tag in the title should be fine.
In your case having one commit for implementing the directive sounds fine.
Also, you need to add a change entry in `Changes` in the `Compiler user-interface and warnings` subsections.

File: toplevel/topdirs.ml [hunk 3]
```diff
@@ -549,11 +549,36 @@ let () =
...
+         | _ -> raise Not_found
+       in
+       let type_decl = Env.find_type path env in
+       let type_id =
+         match path with
+         | Pident type_id -> type_id
+         | Pdot (_, s) -> Ident.create_local s
+         | Papply _ -> raise Not_found
+       in
+       [ Sig_type (type_id, type_decl, Trec_not, Exported) ]
```
Comment (yallop) [inline toplevel/topdirs.ml]: Unconditionally using `Trec_not` here will print type definitions with `nonrec`, even if the types are defined recursively.

File: toplevel/topdirs.ml (hunk 1 above)
Comment (wyn) [reply toplevel/topdirs.ml:571]: i think since this is my first change I'lll stick with the simpler option of adding a new show_constructor function with comments, thanks for the help.

File: toplevel/topdirs.ml (hunk 2 above)
Comment (wyn) [reply toplevel/topdirs.ml]: Ah ok, i see that create_persistent doesn't modify the currentstamp so has no side-effects, that seems like a better approach than create_local.

File: toplevel/topdirs.ml (hunk 3 above)
Comment (wyn) [reply toplevel/topdirs.ml]: thanks, I'll look into how to fix that too.  Looks like I can create another bug report for show_type as that unconditionally uses Trec_not too?

Comment (yallop) [reply toplevel/topdirs.ml]: > Looks like I can create another bug report for show_type as that unconditionally uses `Trec_not` too?

That's a good idea (assuming there isn't one already).

Comment (wyn) [reply toplevel/topdirs.ml]: So my current plan of attack for this bit is to try and extract out the rec_flag (Recursive/Nonrecursive) from the Parsetree.attributes.attr_payload of my 
```
(desc : constructor_description).cstr_attributes
```
Is that the right way to go or am I missing a trick?

My next question is then how to traverse around on the parsetree...

Comment (wyn) [reply toplevel/topdirs.ml]: also can I just clarify what we'd like to see for this last nonrec piece.  For something like:
```
type foo = Foo of bar
and bar = Bar of foo
```

we'd want to see
```
# #show Foo;;
type rec foo = Foo of bar
```

then people can dig further with showing foo/bar/Bar etc. rather than listing out all the and cases at once).

Obviously for non-recursive things it will work as is:
```
# #show Some
type nonrec 'a option = None | Some of 'a
```

Comment (wyn) [reply toplevel/topdirs.ml]: and I think I am missing a trick with the parsetree.attributes.

looking more it seems that a good starting point would be Typedecl.check_recursion function?

Comment (Octachron) [reply toplevel/topdirs.ml]: You don't need to fix this issue in this PR. I think it is fine to keep `Trec_not` for now, in order to be consistent with t`show_type`.  Another option would be to switch to Trec_first` for both. Printing with the right recursivity qualifier is a subject for another PR, and one quite more complex than this one.

Comment (gasche) [reply toplevel/topdirs.ml]: It looks unreasonably difficult to me to access this information from a `Types.type_declaration` value. I'm not super-familiar with this part of the codebase, but it looks like `Types.type_declaration` lacks a lot of useful surface-level information compared to what is available in the `Typedtree` (and in fact ocamldoc reimplements its own environment-gathering logic from the full typedtree in `Odoc_ast.Typedtree_search`).

In your stead I would cut my losses by just using `Trec_first` instead of `Trec_not` bluntly, which *should* result in `nonrec` not being printed. It's a safer default.

**Review by gasche — APPROVED** [review]
This looks good to me, and ready to merge. Would you rebase on top of trunk (there is `Changes` conflict) and merge your commits into a single commit?

**Comment by alainfrisch** [comment]
What about extension constructors (except exceptions)?  Showing the type definition might not give the most useful information for them (the definition of the constructor itself).

**Comment by gasche** [comment]
I think that it's fine to perform step-wise improvements to `#show`, and my impression is that the submitter-pain threshold has been reached on this one, so we could merge now and leave extension constructors to a later PR.

**Comment by gasche** [comment]
Note: here is what an extension-constructors test would look like

```ocaml
type extensible = ..;;
type extensible += A | B;;
{|%%expect{|
type extensible = ..
type extensible += A | B
|}];;

#show A;;
[%%expect {|
|}];;
```

**Comment by alainfrisch** [comment]
Fair enough, but shouldn't we just not show the type definition, i.e. raise Not_found not just for the `exn` type, but also for any extension constructor?

**Comment by gasche** [comment]
@wyn there is a `check-typo` failure:

```
./testsuite/tests/tool-toplevel/show.ml:74.1: [white-at-eof] empty line(s) at EOF
```

could you fix the issue and rebase?

You should also decide what to make of @alainfrisch's proposal to either support extensible variant types, or reject them explicitly. (What does your current branch give on those extensible variants?) I'm happy to merge as-is or wait for further work on this.

**Comment by wyn** [comment]
Yep, sorry about the whitespace, I've just pushed up the rebase for that.

re extensible types, on this branch it behaves like:
```
# type e1 = ..;;
type e1 = ..
# type e1 += A|B;;
type e1 += A | B
# #show A;;
type e1 = ..
```

so I think it is prob best to explicitly not deal with them, and I can raise another PR to get that correct too - if it is another junior-job level task I'm happy to do that PR next - sounds like it probably is.

**Comment by gasche** [comment]
Yes, that sounds good.

Note: it turned out that the present task was actually not so much what I would consider a "junior-level task" -- many inconvenient situations were encountered during the patch writing and discussion/review, due to debatable pre-existing API choices. I don't expect other frontend-related tasks to be noticeably worse.

**Comment by wyn** [comment]
so I just pushed up a change that explicitly does not deal with extension constructors:

```
# type e = ..;;
type e = ..
# type e += A|B;;
type e += A | B
# #show A;;
Unknown element.
```

I'll raise a few more issues off the back of this work:
* fix show directive recursion info - bugfix
* fix show_constructor and show_type for extensible variants - bugfix
* split show directive into one directive for showing type info and one for showing value constructor info? - feature request
* any more...?

**Comment by gasche** [comment]
I'm curious of @alainfrisch opinion, but in retrospect I think that `type e = ..` is a better result than `Unknown element` in this case. (None are great, but the second seems to suggest that the constructor is un-known, which is worse than at least indicating to which extensible type it belongs.)

**Comment by alainfrisch** [comment]
Yeah, perhaps. Showing the type definition brings some info at least, but it's also a bit confusing to show a type (without any constructor) when asking about a constructor. Anyway, this is a minor issue, especially if working on properly showing extension constructors is planned next. Do as you prefer!

**Comment by wyn** [comment]
I'm happy to spend some more time looking into this last bit if people can stand some more reviewing?

**Comment by wyn** [comment]
sorry for all the force-pushes, anyway I've just included a fix for extensible constructors - I see now what people were referring to at the start of this PR, anyway the behaviour is now:

```
# type e = ..;;
type e = ..
# type e += A|B of int;;
type e += A | B of int
# #show e;;
type e = ..
# #show A;;
type e += A
# #show B;;
type e += B of int
```
I've kept the show_type directive the same, the new functionality is for the value constructors.

the previous work remains the same.

**Comment by alainfrisch** [comment]
That's great.  Any reason not to show extension constructors in `#show_constructor`, instead of creating a dedicated entry point?

**Comment by gasche** [comment]
If I read it correctly, tour patch removes the `#show_exception` (sub)directive and adds a `#show_extension_constructor` instead. I would rather keep `#show_exception` exactly as it was before (this preserves compatibility), and add the nicer printing logic to `#show_constructor`, as suggested by @alainfrisch. We could think of having a dedicated show-extension-constructor case in addition, but I don't think it would work well with the the current aggregate-all-results logic of `#show` -- which bites us if two show functions produced overlapping results.

**Comment by wyn** [comment]
@alainfrisch - no not really, probably just not seeing the wood for the trees.  I'll tidy it up a bit.  I think having one longer function that does all constructors is probably better than two functions that use exceptions to signal to each other.

**Comment by wyn** [comment]
@gasche ah ok, good point about compatibility.

So if I revert back the show_exception subdirective, then put the extension logic in show_constructor - is that a good place to stop for this PR.  We can look to improve further in subsequent ones?

File: toplevel/topdirs.ml [hunk 4]
```diff
@@ -545,15 +545,66 @@ let () =
...
+       if is_exception_constructor env desc.cstr_res then
+         raise Not_found;
+       let path =
+         match Ctype.repr desc.cstr_res with
+         | {desc=Tconstr(path, _, _)} -> path
+         | _ -> raise Not_found
+       in
+       if is_extension_constructor desc.cstr_tag then
+         let ret_type =
+           if desc.cstr_generalized then Some Predef.type_exn
```
Comment (gasche) [inline toplevel/topdirs.ml]: I don't think that this is right -- you use `Predef.type_exn` for all extension constructors. You can (you should) test with

```
# type 'a t = ..;;
# type _ t += A : int t;;
[%%expect{|
type 'a t = ..
type _ t += A : int t
|}];;
# #show A;;
[%%expect{|
|}];;
```

I'm not completely sure but I think that `desc.cstr_res` should be usable here.

**Comment by gasche** [comment]
Yes, I think the new point in the feature space is very nice, and we could and should merge the PR as soon as the review process finishes again.

Personally I think that the leftover questions are probably not worth spending someone's time on. They sound like a lot of work for a little gain, with the easier and more convenient part already done by the present PR. Unless you have a fascination for the `#show` directive (who are we to judge?), if you want to contribute further, I would recommend looking at other issues -- and you don't have to limit yourself to "junior" tasks.

File: toplevel/topdirs.ml (hunk 4 above)
Comment (wyn) [reply toplevel/topdirs.ml]: i think you are correct, cstr_res is the type_expr of the result if I understand correctly, also I just noticed I need to be passing in the correct type parameters too, which live on the type_declaration - and with that we get:
```
# type 'a t = ..;;
# type _ t += A : int t;;
# #show A;;
type 'a t += A : int t
```
which seems right, if you pass in the empty list for the type params it gives:
```
type t += A : int t
```
which looks confusing

**Comment by wyn** [comment]
I think we are getting close, if people are happy with this last bit I can squash again and force push the final patch.  

Thanks for all the help everyone.

**Comment by gasche** [comment]
Merged, thanks!

**Comment by wyn** [comment]
Cool, my first PR.

I'll raise those tasks as new issues just so that it is documented somewhere.

File: Changes [hunk 5]
```diff
@@ -30,6 +30,10 @@ Working version
 - #8938: Extend ocamlopt option "-stop-after" to handle "scheduling" argument.
   (Greta Yorsh, review by Florian Angeletti and Sébastien Hinderer)
 
+- #8945: Fix toplevel show directive to work with constructors
```
Comment (SkySkimmer) [inline Changes:33]: Shouldn't the PR itself be mentioned too?

Comment (gasche) [reply Changes:33]: Thanks, I fixed it in trunk.

Comment (wyn) [reply Changes:33]: Sorry about that, so typically one would refer to the PR rather than the original issue?

Comment (gasche) [reply Changes:33]: One refers to both, which lets people find all the discussion about the change.
